File: src/bfs/B_1600.java
package bfs;

import java.util.*;
import java.io.*;

public class B_1600 {
	// 0 ~ 3 : 상하좌우
	// 4 ~ 11 : 말의 이동
	static int[][] dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1},
			{-2, -1}, {-2, 1}, {-1, -2}, {-1, 2},
			{1, -2}, {1, 2}, {2, -1}, {2, 1}};
	static int K, W, H;
	static int[][] arr;
	static int[][] visit;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		K = Integer.parseInt(br.readLine());
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		W = Integer.parseInt(st.nextToken());
		H = Integer.parseInt(st.nextToken());
		arr = new int[H][W];
		visit = new int[H][W];
		
		for (int i=0;i<H;i++) {
			st = new StringTokenizer(br.readLine());
			for (int j=0;j<W;j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
				visit[i][j] = -1;
			}
		}
		
		System.out.println(BFS());
	}
	
	static int BFS() {
		Queue<Monkey> q = new LinkedList<>();
		
		visit[0][0] = K;
		q.offer(new Monkey(0, 0, K, 0));
		
		while (!q.isEmpty()) {
			Monkey now = q.poll();
			
			if (now.r == H - 1 && now.c == W - 1) {
				return now.cnt;
			}
			
			for (int i=0;i<12;i++) {
				int nextR = now.r + dir[i][0];
				int nextC = now.c + dir[i][1];
				
				int nextK = now.k;
				if (i > 3) nextK--;
				
				if (!check(nextR, nextC) || nextK < 0
						|| visit[nextR][nextC] >= nextK || arr[nextR][nextC] == 1) continue;
				
				visit[nextR][nextC] = nextK;
				q.offer(new Monkey(nextR, nextC, nextK, now.cnt + 1));
			}
		}
		
		return -1;
	}
	
	static boolean check(int r, int c) {
		if (r < 0 || r >= H || c < 0 || c >= W) return false;
		return true;
	}
	
	static class Monkey {
		int r, c, k, cnt;
		
		Monkey(int r, int c, int k, int cnt) {
			this.r = r;
			this.c = c;
			this.k = k;
			this.cnt = cnt;
		}
	}

}

File: src/kotlin/bfs/B_1600.kt
import java.util.LinkedList
import kotlin.math.min

var K = 0
var N = 0
var M = 0
lateinit var arr: Array<IntArray>
lateinit var visit: Array<Array<IntArray>>
const val max = Int.MAX_VALUE
var answer = max
val horseX = intArrayOf(-1, -2, -2, -1, 1, 2, 2, 1)
val horseY = intArrayOf(-2, -1, 1, 2, -2, -1, 1, 2)
val dx = intArrayOf(-1, 1, 0, 0)
val dy = intArrayOf(0, 0, -1, 1)

fun main() = with(System.`in`.bufferedReader()) {
    K = readLine().toInt()
    val (m, n) = readLine().split(" ").map { it.toInt() }
    N = n
    M = m
    arr = Array(N) { IntArray(M) }
    visit = Array(N) { Array(M) { IntArray(K + 1) { max } } }

    repeat(N) {
        arr[it] = readLine().split(" ").map { it.toInt() }.toIntArray()
    }

    bfs(0, 0)
    println(if (answer == max) -1 else answer)
}

fun bfs(startX: Int, startY: Int) {
    val q = LinkedList<Monkey>()
    q.offer(Monkey(startX, startY, 0, K))
    visit[startX][startY][K] = 0

    while (!q.isEmpty()) {
        val now = q.poll()

        if (now.x == N - 1 && now.y == M - 1) {
            answer = min(answer, now.distance)
            break
        }

        if (now.k > 0) {
            repeat(8) {
                val nextX = now.x + horseX[it]
                val nextY = now.y + horseY[it]

                if (!check(nextX, nextY)) return@repeat
                if (arr[nextX][nextY] == 1) return@repeat
                if (visit[nextX][nextY][now.k - 1] <= now.distance + 1) return@repeat
                visit[nextX][nextY][now.k - 1] = now.distance + 1
                q.offer(Monkey(nextX, nextY, now.distance + 1, now.k - 1))
            }
        }
        repeat(4) {
            val nextX = now.x + dx[it]
            val nextY = now.y + dy[it]

            if (!check(nextX, nextY)) return@repeat
            if (arr[nextX][nextY] == 1) return@repeat
            if (visit[nextX][nextY][now.k] <= now.distance + 1) return@repeat
            visit[nextX][nextY][now.k] = now.distance + 1
            q.offer(Monkey(nextX, nextY, now.distance + 1, now.k))
        }
    }
}

class Monkey(val x: Int, val y: Int, val distance: Int, val k: Int)

fun check(x: Int, y: Int) = x in 0..<N && y in 0..<M
